
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Documentation &#8212; Structural Motifs 0.1 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Examples" href="examples.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="examples.html" title="Examples"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Structural Motifs 0.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-strum">
<span id="documentation"></span><h1>Documentation<a class="headerlink" href="#module-strum" title="Permalink to this headline">¶</a></h1>
<div class="section" id="strum-structural-motifs">
<h2>StruM: Structural Motifs<a class="headerlink" href="#strum-structural-motifs" title="Permalink to this headline">¶</a></h2>
<p>This package provides functionality for computing structural 
representations of DNA sequence motifs. Estimates for DNA structure
comes from the DiNucleotide Property Database 
(<a class="reference external" href="http://diprodb.leibniz-fli.de/">http://diprodb.leibniz-fli.de/</a>).</p>
<dl class="class">
<dt id="strum.FastStruM">
<em class="property">class </em><code class="descclassname">strum.</code><code class="descname">FastStruM</code><span class="sig-paren">(</span><em>mode='full'</em>, <em>n_process=1</em>, <em>custom_filter=[]</em>, <em>func=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM" title="Permalink to this definition">¶</a></dt>
<dd><p>Differs from the standard StruM in that it cannot
incorporate additional features, it requires all 
sequences to have the same length (or else will 
filter to the shortest length), and uses heuristics
to speed up scoring.</p>
<dl class="method">
<dt id="strum.FastStruM.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>mode='full'</em>, <em>n_process=1</em>, <em>custom_filter=[]</em>, <em>func=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a FastStruM object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>mode</strong> (<em>str.</em>) – <p>Defines which subset of available 
features in the DiProDB table to use. Choose from: 
[‘basic’, ‘protein’, ‘groove’, ‘proteingroove’, ‘unique’, 
‘full’, ‘nucs’, ‘custom’]</p>
<ul>
<li>basic – Twist, Rise, Bend.</li>
<li>protein – (for DNA-protein complex) Roll, Twist, 
Tilt, Slide, Shift, Rise.</li>
<li>groove – Major Groove Width, Major Groove Depth, 
Major Groove Size, Major Groove Distance, Minor 
Groove Width, Minor Groove Depth, Minor Groove Size, 
Minor Groove Distance.</li>
<li>proteingroove – The union of the “protein” and 
“groove” filters.</li>
<li>unique – Filters the table for the first occurrence
of each type of feature.</li>
<li>full – All available DNA features.</li>
<li>nucs – Adenine content, Guanine content, Cytosine 
content, Thymine content.</li>
<li>custom – Manually select desired features.</li>
</ul>
</li>
<li><strong>custom_filter</strong> (<em>list of ints.</em>) – Specifies the indices of 
desired features from the DiProDB table.</li>
<li><strong>n_process</strong> (<em>int.</em>) – Number of threads to use. <code class="docutils literal"><span class="pre">-1</span></code> 
uses all processers.</li>
<li><strong>func</strong> (<em>function.</em>) – Additional scoring functions to incorporate.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.calc_z">
<code class="descname">calc_z</code><span class="sig-paren">(</span><em>x</em>, <em>mu</em>, <em>var</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.calc_z" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate Z-scores for values based on mean and standard deviation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>x</strong> (<em>float</em><em>, </em><em>numpy array.</em>) – Value or values of interest.</li>
<li><strong>mu</strong> (float, numpy array (<code class="docutils literal"><span class="pre">x.shape[1]</span></code>).) – Average of population from which <code class="docutils literal"><span class="pre">x</span></code> was sampled.</li>
<li><strong>var</strong> (float, numpy array (<code class="docutils literal"><span class="pre">x.shape[1]</span></code>).) – Variance of population from which <code class="docutils literal"><span class="pre">x</span></code> was sampled.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The Z-score for all values in <code class="docutils literal"><span class="pre">x</span></code>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">type(x)</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.define_PWM">
<code class="descname">define_PWM</code><span class="sig-paren">(</span><em>seqs</em>, <em>weights=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.define_PWM" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes a position weight matrix from sequences used 
to train the StruM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seqs</strong> (<em>list of str.</em>) – Training set, composed of gapless 
alignment of binding sites of equal length.</li>
<li><strong>weights</strong> (<em>1D array of floats.</em>) – Weights to associate with each of the sequences
in <code class="docutils literal"><span class="pre">seqs</span></code> to use in learning the motif.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None. Sets the position weight matrix <code class="docutils literal"><span class="pre">self.PWM</span></code> 
based on the weighted sequences.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.eval">
<code class="descname">eval</code><span class="sig-paren">(</span><em>struc_kmer</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.eval" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares the structural representation of a sequence 
to the StruM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>struc_kmer</strong> (output of <a class="reference internal" href="#strum.FastStruM.translate" title="strum.FastStruM.translate"><code class="xref py py-func docutils literal"><span class="pre">translate()</span></code></a>.) – A kmer that has been translated to 
structure-space via <a class="reference internal" href="#strum.FastStruM.translate" title="strum.FastStruM.translate"><code class="xref py py-func docutils literal"><span class="pre">translate()</span></code></a>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><em>log</em> score of similarity of kmer to StruM.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">float.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.norm_p">
<code class="descname">norm_p</code><span class="sig-paren">(</span><em>x</em>, <em>mu</em>, <em>var</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.norm_p" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the one-tail p-value for values from the standard normal 
distribution. Adds a ‘pseudocount’ of 10e-300 to avoid underflow.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>x</strong> (<em>float</em><em>, </em><em>numpy array.</em>) – Value or values of interest.</li>
<li><strong>mu</strong> (float, numpy array (<code class="docutils literal"><span class="pre">x.shape[1]</span></code>).) – Average of population from which <code class="docutils literal"><span class="pre">x</span></code> was sampled.</li>
<li><strong>var</strong> (float, numpy array (<code class="docutils literal"><span class="pre">x.shape[1]</span></code>).) – Variance of population from which <code class="docutils literal"><span class="pre">x</span></code> was sampled.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The p-value for all values in <code class="docutils literal"><span class="pre">x</span></code> relative to <code class="docutils literal"><span class="pre">mu</span></code>
and <code class="docutils literal"><span class="pre">var</span></code>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">type(x)</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.print_PWM">
<code class="descname">print_PWM</code><span class="sig-paren">(</span><em>labels=False</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.print_PWM" title="Permalink to this definition">¶</a></dt>
<dd><p>Pretty prints the PWM to std_out.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>bool.</em>) – Flag indicating whether to print the PWM
with labels indicating the position associated with 
each column, and the nucleotide associated with each
row.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Formatted position weight matrix suitable for
display, or use in the MEME suite, e.g. Also prints
the PWM to <code class="docutils literal"><span class="pre">std_out</span></code>.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">str.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.read_FASTA">
<code class="descname">read_FASTA</code><span class="sig-paren">(</span><em>fasta_file</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.read_FASTA" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a FASTA formatted file for headers and sequences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fasta_file</strong> – FASTA formatted file containing DNA sequences.</td>
</tr>
<tr class="field-even field"><th class="field-name">Type:</th><td class="field-body">file object</td>
</tr>
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The headers and sequences from the FASTA file, as two 
separate lists.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">(list, list)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.rev_comp">
<code class="descname">rev_comp</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.rev_comp" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse complement (uppercase) DNA sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>seq</strong> (<em>str.</em>) – DNA sequence, all uppercase characters,
composed of letters from set ACGTN.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Reverse complement of <code class="docutils literal"><span class="pre">seq</span></code>.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">str.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.score_seq">
<code class="descname">score_seq</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.score_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Scores a sequence using pre-calculated StruM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>seq</strong> (<em>str.</em>) – DNA sequence, all uppercase characters,
composed of letters from set ACGTN.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Vector of scores for similarity of each kmer
in <code class="docutils literal"><span class="pre">seq</span></code> to the StruM.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">1D array.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.train">
<code class="descname">train</code><span class="sig-paren">(</span><em>training_sequences</em>, <em>weights=None</em>, <em>lim=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.train" title="Permalink to this definition">¶</a></dt>
<dd><p>Learn structural motif from a set of known binding sites.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>training_sequences</strong> (<em>list of str.</em>) – Training set, composed of gapless 
alignment of binding sites of equal length.</li>
<li><strong>weights</strong> (<em>1D array of floats.</em>) – Weights to associate with each of the sequences
in <code class="docutils literal"><span class="pre">training_sequences</span></code> to use in learning the motif.</li>
<li><strong>lim</strong> (<em>float</em>) – Minimum value allowed for variation in a given
position-specific-feature. Useful to prevent <em>any</em>
deviation at that position from resulting in a probability of
0.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None. Defines the structural motif <code class="docutils literal"><span class="pre">self.strum</span></code> and the
corresponding position weight matrix <code class="docutils literal"><span class="pre">self.PWM</span></code>.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.train_EM">
<code class="descname">train_EM</code><span class="sig-paren">(</span><em>data</em>, <em>fasta=True</em>, <em>params=None</em>, <em>k=10</em>, <em>max_iter=1000</em>, <em>convergence_criterion=0.001</em>, <em>random_seed=0</em>, <em>n_init=1</em>, <em>lim=None</em>, <em>seqlength=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.train_EM" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs Expectation-Maximization on a set of sequences 
to find motif.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<em>list of str</em><em>, </em><em>open file object referring to a 
FASTA file.</em>) – A set of sequences to use for training the model.
Assumed to have one occurrence of the binding site per 
sequence.</li>
<li><strong>fasta</strong> (<em>bool.</em>) – Flag indicating whether <code class="docutils literal"><span class="pre">data</span></code> points to an
open file object containing a FASTA formatted file with
DNA sequences.</li>
<li><strong>params</strong> (<code class="docutils literal"><span class="pre">*args</span></code>, <code class="docutils literal"><span class="pre">**kwargs</span></code>.) – Additional parameters to pass to <code class="docutils literal"><span class="pre">self.func</span></code>,
if defined.</li>
<li><strong>k</strong> (<em>int.</em>) – Size of binding site to consider. Since dinucleotides
are considered, in sequence-space the size of the binding
site will be <code class="docutils literal"><span class="pre">k</span> <span class="pre">+</span> <span class="pre">1</span></code>.</li>
<li><strong>max_iter</strong> (<em>int.</em>) – Maximum number of iterations of Expecation
Maximization to perform if convergence is not attained.</li>
<li><strong>convergence_criterion</strong> (<em>float.</em>) – If the change in the likelihood
between two iterations is less than this value, the model
is considered to have converged.</li>
<li><strong>random_seed</strong> (<em>int.</em>) – Seed for the random number generator used
in the EM algorithm for initialization.</li>
<li><strong>n_init</strong> (<em>int.</em>) – Number of random restarts of the EM algorithm
to perform.</li>
<li><strong>lim</strong> (<em>float</em>) – Minimum value allowed for variation in a given
position-specific-feature. Useful to prevent <em>any</em>
deviation at that position from resulting in a probability of
0.</li>
<li><strong>seqlength</strong> (<em>int.</em>) – If set, the sequences in the training data
will be trimmed symmetrically to this length. 
.. note:: This must be longer than the shortes sequence.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None. Defines the structural motif <code class="docutils literal"><span class="pre">self.strum</span></code> and the
corresponding position weight matrix <code class="docutils literal"><span class="pre">self.PWM</span></code>.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.FastStruM.translate">
<code class="descname">translate</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.FastStruM.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert sequence from string to structural representation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>seq</strong> (<em>str.</em>) – DNA sequence, all uppercase characters,
composed of letters from set ACGTN.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Sequence in structural representation.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">1D numpy array of floats.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="strum.read_diprodb">
<code class="descclassname">strum.</code><code class="descname">read_diprodb</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#strum.read_diprodb" title="Permalink to this definition">¶</a></dt>
<dd><p>Load the values from the DiNucleotide Property Database as a lookup table.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="examples.html"
                        title="previous chapter">Examples</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="examples.html" title="Examples"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Structural Motifs 0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, Peter DeFord.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.6.
    </div>
  </body>
</html>