

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Documentation &mdash; Structural Motifs 0.2 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="Structural Motifs 0.2 documentation" href="index.html"/>
        <link rel="prev" title="Examples" href="examples.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Structural Motifs
          

          
          </a>

          
            
            
              <div class="version">
                0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction to Structural Motifs</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#strum-structural-motifs">StruM: Structural Motifs</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Structural Motifs</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Documentation</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/code.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-strum">
<span id="documentation"></span><h1>Documentation<a class="headerlink" href="#module-strum" title="Permalink to this headline">¶</a></h1>
<div class="section" id="strum-structural-motifs">
<h2>StruM: Structural Motifs<a class="headerlink" href="#strum-structural-motifs" title="Permalink to this headline">¶</a></h2>
<p>This package provides functionality for computing structural 
representations of DNA sequence motifs. Estimates for DNA structure
comes from the DiNucleotide Property Database 
(<a class="reference external" href="http://diprodb.leibniz-fli.de/">http://diprodb.leibniz-fli.de/</a>).</p>
<dl class="class">
<dt id="strum.StruM">
<em class="property">class </em><code class="descclassname">strum.</code><code class="descname">StruM</code><span class="sig-paren">(</span><em>mode='full'</em>, <em>n_process=1</em>, <em>custom_filter=[]</em>, <em>func=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM" title="Permalink to this definition">¶</a></dt>
<dd><p>StruM Object: Train and work with Structural Motifs</p>
<p>StruMs can be learned via maximum likelihood (<a class="reference internal" href="#strum.StruM.train" title="strum.StruM.train"><code class="xref py py-func docutils literal"><span class="pre">train()</span></code></a>) from a
known set of aligned binding sites, or via expectation maximization
(<a class="reference internal" href="#strum.StruM.train_EM" title="strum.StruM.train_EM"><code class="xref py py-func docutils literal"><span class="pre">train_EM()</span></code></a>) from a set of broader regions.</p>
<p>For speed, this package will trim the sequences to be the same
length. This can be user specified, else the shortest sequence 
will be used as a guide.</p>
<p>Additional features other than those from DiProDB can be 
incorporated by using the <a class="reference internal" href="#strum.StruM.update" title="strum.StruM.update"><code class="xref py py-func docutils literal"><span class="pre">update()</span></code></a> method.</p>
<dl class="method">
<dt id="strum.StruM.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>mode='full'</em>, <em>n_process=1</em>, <em>custom_filter=[]</em>, <em>func=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a FastStruM object.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>mode</strong> (<em>str.</em>) – <p>Defines which subset of available 
features in the DiProDB table to use. Choose from: 
[‘basic’, ‘protein’, ‘groove’, ‘proteingroove’, ‘unique’, 
‘full’, ‘nucs’, ‘custom’]</p>
<table border="1" class="docutils">
<colgroup>
<col width="26%" />
<col width="74%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">MODE</th>
<th class="head">Features</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>basic</td>
<td>Twist, Rise, Bend.</td>
</tr>
<tr class="row-odd"><td>protein</td>
<td>(for DNA-protein complex) Roll, Twist,
Tilt, Slide, Shift, Rise.</td>
</tr>
<tr class="row-even"><td>groove</td>
<td>Major Groove Width, Major Groove Depth,
Major Groove Size, Major Groove Distance,
Minor Groove Width, Minor Groove Depth,
Minor Groove Size, Minor Groove Distance.</td>
</tr>
<tr class="row-odd"><td>proteingroove</td>
<td>The union of the “protein” and “groove”
filters.</td>
</tr>
<tr class="row-even"><td>unique</td>
<td>Filters the table for the first
occurrence of each type of feature.</td>
</tr>
<tr class="row-odd"><td>full</td>
<td>All available double stranded DNA
features.</td>
</tr>
<tr class="row-even"><td>nucs</td>
<td>Adenine content, Guanine content,
Cytosine content, Thymine content.</td>
</tr>
<tr class="row-odd"><td>custom</td>
<td>Manually select desired features.</td>
</tr>
</tbody>
</table>
</li>
<li><strong>custom_filter</strong> (<em>list of ints.</em>) – Specifies the indices of 
desired features from the DiProDB table.</li>
<li><strong>n_process</strong> (<em>int.</em>) – Number of threads to use. <code class="docutils literal"><span class="pre">-1</span></code> 
uses all processers.</li>
<li><strong>func</strong> (<em>function.</em>) – Additional scoring functions to incorporate.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.calc_z">
<code class="descname">calc_z</code><span class="sig-paren">(</span><em>x</em>, <em>mu</em>, <em>var</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.calc_z" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate Z-scores for values based on mean and standard deviation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>x</strong> (<em>float</em><em>, </em><em>numpy array.</em>) – Value or values of interest.</li>
<li><strong>mu</strong> (float, numpy array (<code class="docutils literal"><span class="pre">x.shape[1]</span></code>).) – Average of population from which <code class="docutils literal"><span class="pre">x</span></code> was sampled.</li>
<li><strong>var</strong> (float, numpy array (<code class="docutils literal"><span class="pre">x.shape[1]</span></code>).) – Variance of population from which <code class="docutils literal"><span class="pre">x</span></code> was sampled.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The Z-score for all values in <code class="docutils literal"><span class="pre">x</span></code>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">type(x)</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.define_PWM">
<code class="descname">define_PWM</code><span class="sig-paren">(</span><em>seqs</em>, <em>weights=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.define_PWM" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes a position weight matrix from sequences used 
to train the StruM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seqs</strong> (<em>list of str.</em>) – Training set, composed of gapless 
alignment of binding sites of equal length.</li>
<li><strong>weights</strong> (<em>1D array of floats.</em>) – Weights to associate with each of the sequences
in <code class="docutils literal"><span class="pre">seqs</span></code> to use in learning the motif.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None. Sets the position weight matrix <code class="docutils literal"><span class="pre">self.PWM</span></code> 
based on the weighted sequences.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.eval">
<code class="descname">eval</code><span class="sig-paren">(</span><em>struc_kmer</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.eval" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares the structural representation of a sequence 
to the StruM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>struc_kmer</strong> (output of <a class="reference internal" href="#strum.StruM.translate" title="strum.StruM.translate"><code class="xref py py-func docutils literal"><span class="pre">translate()</span></code></a>.) – A kmer that has been translated to 
structure-space via <a class="reference internal" href="#strum.StruM.translate" title="strum.StruM.translate"><code class="xref py py-func docutils literal"><span class="pre">translate()</span></code></a>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><em>log</em> score of similarity of kmer to StruM.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">float.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.eval_filt">
<code class="descname">eval_filt</code><span class="sig-paren">(</span><em>struc_kmer</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.eval_filt" title="Permalink to this definition">¶</a></dt>
<dd><p>A variation on <a class="reference internal" href="#strum.StruM.eval" title="strum.StruM.eval"><code class="xref py py-func docutils literal"><span class="pre">eval()</span></code></a> that masks non-specific features.</p>
<p>Once the <cite>self.filter_mask</cite> is generated, this method becomes available. 
This compares the structural representation of a sequence to the 
StruM.</p>
<p>Refer to <a class="reference internal" href="#strum.StruM.eval" title="strum.StruM.eval"><code class="xref py py-func docutils literal"><span class="pre">eval()</span></code></a> for more information about the arguments.</p>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.filter">
<code class="descname">filter</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Update StruM to ignore uninformative position-specific features.</p>
<p>Features with a high variance, i.e. non-specific features, do not
contribute to the specificity of the StruM model. Filtering them
out may increase the signal-to-noise ratio. The position-specific-
features are rank ordered by their variance, and a univariate 
spline is fit to the distribution. The point of inflection is 
used as the threshold for masking less specific features.</p>
<p>Once this method is run and the attribute <cite>self.filter_mask</cite> is 
generated, two additional methods will become available:
<a class="reference internal" href="#strum.StruM.score_seq_filt" title="strum.StruM.score_seq_filt"><code class="xref py py-func docutils literal"><span class="pre">score_seq_filt()</span></code></a> and <a class="reference internal" href="#strum.StruM.eval_filt" title="strum.StruM.eval_filt"><code class="xref py py-func docutils literal"><span class="pre">eval_filt()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.func_translate">
<code class="descname">func_translate</code><span class="sig-paren">(</span><em>seq</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.func_translate" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Convert sequence from string to structural representation,</dt>
<dd>with additional features.</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seq</strong> (<em>(</em><em>str</em><em>, </em><em>[</em><em>args</em><em>]</em><em>)</em><em></em>) – DNA sequence, all uppercase characters,
composed of letters from set ACGTN.</li>
<li><strong>**kwargs</strong> – Additional keyword arguments required
by <code class="docutils literal"><span class="pre">self.func</span></code>.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Sequence in structural representation.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">1D numpy array of floats.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.norm_p">
<code class="descname">norm_p</code><span class="sig-paren">(</span><em>x</em>, <em>mu</em>, <em>var</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.norm_p" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the one-tail p-value for values from the standard normal 
distribution. Adds a ‘pseudocount’ of 10e-300 to avoid underflow.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>x</strong> (<em>float</em><em>, </em><em>numpy array.</em>) – Value or values of interest.</li>
<li><strong>mu</strong> (float, numpy array (<code class="docutils literal"><span class="pre">x.shape[1]</span></code>).) – Average of population from which <code class="docutils literal"><span class="pre">x</span></code> was sampled.</li>
<li><strong>var</strong> (float, numpy array (<code class="docutils literal"><span class="pre">x.shape[1]</span></code>).) – Variance of population from which <code class="docutils literal"><span class="pre">x</span></code> was sampled.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The p-value for all values in <code class="docutils literal"><span class="pre">x</span></code> relative to <code class="docutils literal"><span class="pre">mu</span></code>
and <code class="docutils literal"><span class="pre">var</span></code>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">type(x)</span></code></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.plot">
<code class="descname">plot</code><span class="sig-paren">(</span><em>save_path</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a graphical representation of the StruM.</p>
<p>Generates an image displaying the trends of each of the 
features in the StruM. The line indicates the average 
(scaled) value for that feature at each position. Shading
represents +/- 1 standard devation.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This generates one row per feature. If you are including 
many features, your plot may be excessively tall.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>save_path</strong> (<em>str.</em>) – Filename to use when saving the image.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.print_PWM">
<code class="descname">print_PWM</code><span class="sig-paren">(</span><em>labels=False</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.print_PWM" title="Permalink to this definition">¶</a></dt>
<dd><p>Pretty prints the PWM to std_out.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>bool.</em>) – Flag indicating whether to print the PWM
with labels indicating the position associated with 
each column, and the nucleotide associated with each
row.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Formatted position weight matrix suitable for
display, or use in the MEME suite, e.g. Also prints
the PWM to <code class="docutils literal"><span class="pre">std_out</span></code>.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">str.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.read_FASTA">
<code class="descname">read_FASTA</code><span class="sig-paren">(</span><em>fasta_file</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.read_FASTA" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a FASTA formatted file for headers and sequences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fasta_file</strong> – FASTA formatted file containing DNA sequences.</td>
</tr>
<tr class="field-even field"><th class="field-name">Type:</th><td class="field-body">file object</td>
</tr>
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The headers and sequences from the FASTA file, as two 
separate lists.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">(list, list)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.rev_comp">
<code class="descname">rev_comp</code><span class="sig-paren">(</span><em>seq</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.rev_comp" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse complement (uppercase) DNA sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>seq</strong> (<em>str.</em>) – DNA sequence, all uppercase characters,
composed of letters from set ACGTN.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Reverse complement of <code class="docutils literal"><span class="pre">seq</span></code>.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">str.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.score_seq">
<code class="descname">score_seq</code><span class="sig-paren">(</span><em>seq</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.score_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Scores a sequence using pre-calculated StruM.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>seq</strong> (<em>str.</em>) – DNA sequence, all uppercase characters,
composed of letters from set ACGTN.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Vector of scores for similarity of each kmer
in <code class="docutils literal"><span class="pre">seq</span></code> to the StruM.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">1D array.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.score_seq_filt">
<code class="descname">score_seq_filt</code><span class="sig-paren">(</span><em>seq</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.score_seq_filt" title="Permalink to this definition">¶</a></dt>
<dd><p>A variation on <a class="reference internal" href="#strum.StruM.score_seq" title="strum.StruM.score_seq"><code class="xref py py-func docutils literal"><span class="pre">score_seq()</span></code></a> that masks non-specific features.</p>
<p>Once the <cite>self.filter_mask</cite> is generated, this method becomes available. 
This scores a sequence with the precomputed StruM, masking non-specific
features.</p>
<p>Refer to <a class="reference internal" href="#strum.StruM.score_seq" title="strum.StruM.score_seq"><code class="xref py py-func docutils literal"><span class="pre">score_seq()</span></code></a> for more information about the arguments.</p>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.train">
<code class="descname">train</code><span class="sig-paren">(</span><em>training_sequences</em>, <em>weights=None</em>, <em>lim=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.train" title="Permalink to this definition">¶</a></dt>
<dd><p>Learn structural motif from a set of known binding sites.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>training_sequences</strong> (<em>list of str. Or if updated</em><em>, </em><em>list of tuples:</em><em> (</em><em>str</em><em>, </em><em>[</em><em>args</em><em>]</em><em>)</em>) – Training set, composed of gapless 
alignment of binding sites of equal length.</li>
<li><strong>weights</strong> (<em>1D array of floats.</em>) – Weights to associate with each of the sequences
in <code class="docutils literal"><span class="pre">training_sequences</span></code> to use in learning the motif.</li>
<li><strong>lim</strong> (<em>float</em>) – Minimum value allowed for variation in a given
position-specific-feature. Useful to prevent <em>any</em>
deviation at that position from resulting in a probability of
0.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None. Defines the structural motif <code class="docutils literal"><span class="pre">self.strum</span></code> and the
corresponding position weight matrix <code class="docutils literal"><span class="pre">self.PWM</span></code>.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.train_EM">
<code class="descname">train_EM</code><span class="sig-paren">(</span><em>data</em>, <em>fasta=True</em>, <em>params=None</em>, <em>k=10</em>, <em>max_iter=1000</em>, <em>convergence_criterion=0.001</em>, <em>random_seed=0</em>, <em>n_init=1</em>, <em>lim=None</em>, <em>seqlength=None</em>, <em>background=None</em>, <em>seed_motif=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.train_EM" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs Expectation-Maximization on a set of sequences 
to find motif.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<em>list of str</em><em>, </em><em>open file object referring to a 
FASTA file.</em>) – A set of sequences to use for training the model.
Assumed to have one occurrence of the binding site per 
sequence.</li>
<li><strong>fasta</strong> (<em>bool.</em>) – Flag indicating whether <code class="docutils literal"><span class="pre">data</span></code> points to an
open file object containing a FASTA formatted file with
DNA sequences.</li>
<li><strong>params</strong> (<code class="docutils literal"><span class="pre">*args</span></code>, <code class="docutils literal"><span class="pre">**kwargs</span></code>.) – Additional parameters to pass to <code class="docutils literal"><span class="pre">self.func</span></code>,
if defined.</li>
<li><strong>k</strong> (<em>int.</em>) – Size of binding site to consider. Since dinucleotides
are considered, in sequence-space the size of the binding
site will be <code class="docutils literal"><span class="pre">k</span> <span class="pre">+</span> <span class="pre">1</span></code>.</li>
<li><strong>max_iter</strong> (<em>int.</em>) – Maximum number of iterations of Expecation
Maximization to perform if convergence is not attained.</li>
<li><strong>convergence_criterion</strong> (<em>float.</em>) – If the change in the likelihood
between two iterations is less than this value, the model
is considered to have converged.</li>
<li><strong>random_seed</strong> (<em>int.</em>) – Seed for the random number generator used
in the EM algorithm for initialization.</li>
<li><strong>n_init</strong> (<em>int.</em>) – Number of random restarts of the EM algorithm
to perform.</li>
<li><strong>lim</strong> (<em>float</em>) – Minimum value allowed for variation in a given
position-specific-feature. Useful to prevent <em>any</em>
deviation at that position from resulting in a probability of
0.</li>
<li><strong>seqlength</strong> (<em>int.</em>) – If set, the sequences in the training data
will be trimmed symmetrically to this length. 
.. note:: This must be longer than the shortes sequence.</li>
<li><strong>background</strong> (<em>str.</em>) – <p>Method to use for computing the background
model. Default is to assume equal probability of each
dinucleotide. Passing <code class="docutils literal"><span class="pre">&quot;compute&quot;</span></code> adapts the background
model to represent the dinucleotide frequencies in the
training sequences. Otherwise this can the path to a tab 
delimited file specifying the representation of each 
dinucleotide. E.g.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">AA</span>      <span class="mf">0.0625</span>
<span class="n">AC</span>      <span class="mf">0.0625</span>
<span class="o">...</span>
</pre></div>
</div>
</li>
<li><strong>seed_motif</strong> (<code class="docutils literal"><span class="pre">StruM.strum</span></code>.) – Optional. A <cite>StruM.strum</cite> to use for 
initializing the Expectation-Maximization algorithm.
If set, <code class="docutils literal"><span class="pre">k</span></code> will be replaced by the corresponding
value from the <code class="docutils literal"><span class="pre">seed_motif</span></code>’s shape. <code class="docutils literal"><span class="pre">n_init</span></code> will
also be reset to 1.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">None. Defines the structural motif <code class="docutils literal"><span class="pre">self.strum</span></code> and the
corresponding position weight matrix <code class="docutils literal"><span class="pre">self.PWM</span></code>.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.translate">
<code class="descname">translate</code><span class="sig-paren">(</span><em>seq</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert sequence from string to structural representation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seq</strong> (<em>str.</em>) – DNA sequence, all uppercase characters,
composed of letters from set ACGTN.</li>
<li><strong>**kwargs</strong> – Ignored</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Sequence in structural representation.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">1D numpy array of floats.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="strum.StruM.update">
<code class="descname">update</code><span class="sig-paren">(</span><em>features</em>, <em>func</em>, <em>data=None</em><span class="sig-paren">)</span><a class="headerlink" href="#strum.StruM.update" title="Permalink to this definition">¶</a></dt>
<dd><p>Update the StruM to incorporate additional features.</p>
<p>Using this method will change the behavior of other methods,
especially the <a class="reference internal" href="#strum.StruM.translate" title="strum.StruM.translate"><code class="xref py py-func docutils literal"><span class="pre">translate()</span></code></a> method is replaced by 
<a class="reference internal" href="#strum.StruM.func_translate" title="strum.StruM.func_translate"><code class="xref py py-func docutils literal"><span class="pre">func_translate()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>features</strong> (<em>list of str.</em>) – Text description or label of the feature(s)
being added into the model.</li>
<li><strong>func</strong> (<em>function.</em>) – The scoring function that produces the additional
features. These may be computed on sequence alone, or by
incorporating additional data. The output must be an array
of shape <code class="docutils literal"><span class="pre">[n,</span> <span class="pre">l-1]</span></code>, where <code class="docutils literal"><span class="pre">n</span></code> is the number of additional 
features (<code class="docutils literal"><span class="pre">len(features)</span></code>) and <code class="docutils literal"><span class="pre">l</span></code> is the length of the 
sequence being scored. The first argument of the function must
be the sequence being scored.</li>
<li><strong>data</strong> – Additional data that is used by the new function. May
be a lookup table, for example, or a reference to an outside
file.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="strum.read_diprodb">
<code class="descclassname">strum.</code><code class="descname">read_diprodb</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#strum.read_diprodb" title="Permalink to this definition">¶</a></dt>
<dd><p>Load the values from the DiNucleotide Property Database as a lookup table.</p>
</dd></dl>

</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="examples.html" class="btn btn-neutral" title="Examples" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Peter DeFord.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.2',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>