

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Examples &mdash; Structural Motifs 0.1 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="Structural Motifs 0.1 documentation" href="index.html"/>
        <link rel="next" title="Documentation" href="code.html"/>
        <link rel="prev" title="Installation Instructions" href="installation.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Structural Motifs
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction to Structural Motifs</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation Instructions</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basic-example">Basic Example</a></li>
<li class="toctree-l2"><a class="reference internal" href="#de-novo-motif-finding-example"><em>De Novo</em> Motif Finding Example</a></li>
<li class="toctree-l2"><a class="reference internal" href="#adding-additional-features">Adding Additional Features</a></li>
<li class="toctree-l2"><a class="reference internal" href="#plotting-a-graphical-representation-of-the-strum">Plotting a graphical representation of the StruM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="code.html">Documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Structural Motifs</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Examples</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/examples.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="examples">
<h1><a class="toc-backref" href="#id1">Examples</a><a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<p>This page contains simple examples showcasing the basic functionality
of the StruM package.</p>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#examples" id="id1">Examples</a><ul>
<li><a class="reference internal" href="#basic-example" id="id2">Basic Example</a></li>
<li><a class="reference internal" href="#de-novo-motif-finding-example" id="id3"><em>De Novo</em> Motif Finding Example</a></li>
<li><a class="reference internal" href="#adding-additional-features" id="id4">Adding Additional Features</a></li>
<li><a class="reference internal" href="#plotting-a-graphical-representation-of-the-strum" id="id5">Plotting a graphical representation of the StruM</a></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="basic-example">
<h2><a class="toc-backref" href="#id2">Basic Example</a><a class="headerlink" href="#basic-example" title="Permalink to this headline">¶</a></h2>
<p>The basic usage of the StruM package covers the case where the binding
site is already known. If the PWM is known already, a tool such as <a class="reference external" href="http://meme-suite.org/doc/fimo.html">FIMO</a> could be used to determine
possible matches to the binding site.</p>
<p>In this case, a maximum likelihood StruM can be computed directly from
these sequences. Source code for this example can be found here:
<a class="reference external" href="https://github.com/pdeford/StructuralMotifs/blob/master/examples/basic.py">basic.py</a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env python</span>

<span class="c1"># Imports</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">strum</span> <span class="k">import</span> <span class="n">strum</span>

<span class="c1"># Sequences representing some of the variability of the</span>
<span class="c1"># FOXA1 binding site.</span>
<span class="n">training_sequences</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;CTGTGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTAAGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTGTGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTGTGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CAGAGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTAAGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTGTGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CAATGTAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTGAGTAAATA&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="c1"># Initialize a new StruM object, using the basic features</span>
<span class="c1"># from the DiProDB table.</span>
<span class="n">motif</span> <span class="o">=</span> <span class="n">strum</span><span class="o">.</span><span class="n">StruM</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;basic&#39;</span><span class="p">)</span>

<span class="c1"># Use the training sequences to define the StruM,</span>
<span class="c1"># ensuring that that the variation of all position-specific</span>
<span class="c1"># features is at least 10e-5 (lim)</span>
<span class="n">motif</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">training_sequences</span><span class="p">,</span> <span class="n">lim</span><span class="o">=</span><span class="mi">10</span><span class="o">**-</span><span class="mi">5</span><span class="p">)</span>

<span class="c1"># Print out what the PWM would look like for these sequences</span>
<span class="n">motif</span><span class="o">.</span><span class="n">print_PWM</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Define an example sequence to analyze.</span>
<span class="n">test_sequence</span> <span class="o">=</span> <span class="s2">&quot;ACGTACTGCAGAGCAAACAACTGATCGGATC&quot;</span>
<span class="c1"># Reverse complement it, as the best match may not be on</span>
<span class="c1"># the forward strand.</span>
<span class="n">reversed_test</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">rev_comp</span><span class="p">(</span><span class="n">test_sequence</span><span class="p">)</span>

<span class="c1"># Get a score of the similarity for each kmer in the test</span>
<span class="c1"># sequence to the StruM.</span>
<span class="n">forward_scores</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">score_seq</span><span class="p">(</span><span class="n">test_sequence</span><span class="p">)</span>
<span class="n">reverse_scores</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">score_seq</span><span class="p">(</span><span class="n">reversed_test</span><span class="p">)</span>

<span class="c1"># Find the best match.</span>
<span class="c1">## Determine the strand the best match lies on</span>
<span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">reverse_scores</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">forward_scores</span><span class="p">):</span>
    <span class="n">score</span><span class="p">,</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">reverse_scores</span><span class="p">,</span> <span class="n">reversed_test</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">score</span><span class="p">,</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">forward_scores</span><span class="p">,</span> <span class="n">test_sequence</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;+&quot;</span>
<span class="c1">## Find position in the sequence</span>
<span class="n">best_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>
<span class="n">best_seq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">best_pos</span> <span class="p">:</span> <span class="n">best_pos</span> <span class="o">+</span> <span class="n">motif</span><span class="o">.</span><span class="n">k</span><span class="p">]</span>
<span class="n">best_score</span> <span class="o">=</span> <span class="n">score</span><span class="p">[</span><span class="n">best_pos</span><span class="p">]</span>

<span class="c1"># Print the results to screen</span>
<span class="nb">print</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">best_pos</span><span class="p">,</span> <span class="n">best_seq</span><span class="p">,</span> <span class="n">best_score</span>
</pre></div>
</div>
<p>You should see the position weight matrix that would be derived
from these sequences followed by the strand, position, matching
kmer, and score for the best match in the test sequence to the
StruM. The labels surrounding the PWM can be toggled off by
changing the <code class="docutils literal"><span class="pre">labels</span></code> parameter in <a class="reference internal" href="code.html#strum.StruM.print_PWM" title="strum.StruM.print_PWM"><code class="xref py py-func docutils literal"><span class="pre">strum.StruM.print_PWM()</span></code></a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>    <span class="mi">1</span>     <span class="mi">2</span>     <span class="mi">3</span>     <span class="mi">4</span>     <span class="mi">5</span>     <span class="mi">6</span>     <span class="mi">7</span>     <span class="mi">8</span>     <span class="mi">9</span>    <span class="mi">10</span>    <span class="mi">11</span>
<span class="n">A</span> <span class="mf">0.000</span> <span class="mf">0.222</span> <span class="mf">0.333</span> <span class="mf">0.444</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">1.000</span> <span class="mf">1.000</span> <span class="mf">1.000</span> <span class="mf">0.000</span> <span class="mf">1.000</span>
<span class="n">C</span> <span class="mf">1.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.778</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.889</span> <span class="mf">0.000</span>
<span class="n">G</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.667</span> <span class="mf">0.000</span> <span class="mf">1.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span>
<span class="n">T</span> <span class="mf">0.000</span> <span class="mf">0.778</span> <span class="mf">0.000</span> <span class="mf">0.556</span> <span class="mf">0.000</span> <span class="mf">0.222</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.000</span> <span class="mf">0.111</span> <span class="mf">0.000</span>

<span class="o">+</span> <span class="mi">8</span> <span class="n">CAGAGCAAACA</span> <span class="o">-</span><span class="mf">43.2745526744</span>
</pre></div>
</div>
</div>
<div class="section" id="de-novo-motif-finding-example">
<h2><a class="toc-backref" href="#id3"><em>De Novo</em> Motif Finding Example</a><a class="headerlink" href="#de-novo-motif-finding-example" title="Permalink to this headline">¶</a></h2>
<p>More often, the task is to take a set of sequences and perform <em>de novo</em>
motif finding, simultaneously modeling the motif and identifying the
putative binding sites. The StruM package allows you to do this using
expectation maximization.</p>
<p>The following example uses the EM capabilities of the StruM package to
identify the motif GATTACA randomly inserted into 50 sequences. The
source gode for this example can be downloaded here: <a class="reference external" href="https://github.com/pdeford/StructuralMotifs/blob/master/examples/em.py">em.py</a>.</p>
<p>Refer to <a class="reference internal" href="code.html#strum.StruM.train_EM" title="strum.StruM.train_EM"><code class="xref py py-func docutils literal"><span class="pre">strum.StruM.train_EM()</span></code></a> for more info on the parameters for the EM
algorithm.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env python</span>

<span class="c1"># Imports</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">strum</span> <span class="k">import</span> <span class="n">strum</span>

<span class="c1"># 50 example sequences that contain the sequence GATTACA</span>
<span class="n">training_sequences</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;CGATTACAGATCTCCCGCGACCCTT&quot;</span><span class="p">,</span> <span class="s2">&quot;GATGATTACAAGATGCGTCGAATAT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AGCCCTGTCCGCAGATTACAACCAC&quot;</span><span class="p">,</span> <span class="s2">&quot;AGTTAACTCCCTAGATTACATTTGT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CGCTACAAAGTAAAGGAGATTACAT&quot;</span><span class="p">,</span> <span class="s2">&quot;CTGATTACATCCTGTCCGAGGCGTG&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CGATTACATTCAACTAGATGCGCGC&quot;</span><span class="p">,</span> <span class="s2">&quot;GAACGGCATGGGCGATTACAACACT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CGGGGTGATCGTAATGTGATTACAA&quot;</span><span class="p">,</span> <span class="s2">&quot;TGGTGCCCTGATTACACCTTACATG&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GGCGAAGATTACATCCTCGGCCCAT&quot;</span><span class="p">,</span> <span class="s2">&quot;ATCGGATTACATGTACTCGTCCACG&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TCGGGGATTACAGGGGAGACGCTTA&quot;</span><span class="p">,</span> <span class="s2">&quot;AGGTAGATTACATCGTTTTATTAGT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TATTGTGCTCGATTACAAGCAGGCC&quot;</span><span class="p">,</span> <span class="s2">&quot;CAGACCGCTTACACGTTGATTACAA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GACACCCTCGATTACACCTCGTATA&quot;</span><span class="p">,</span> <span class="s2">&quot;GGAACCGCGCGGATTACACGCGAGA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GTTGATTACAAGGGAAACATACTTG&quot;</span><span class="p">,</span> <span class="s2">&quot;CCCACACATTAGCTCGAGATTACAT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GCAGAGTACCCTGCGGCGATTACAA&quot;</span><span class="p">,</span> <span class="s2">&quot;ATACTCACGCATACAGATTACAAGA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GGTATGCATCGCGATTACAGCACTG&quot;</span><span class="p">,</span> <span class="s2">&quot;GGATTACAGTGAGCCTGCACCTTGA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TTGGATTACATGGCCAAACTCCACT&quot;</span><span class="p">,</span> <span class="s2">&quot;GGCCGGCAGAGATTACACTAGAGAG&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ACAGATTACAGTGCAAATTGAGCAG&quot;</span><span class="p">,</span> <span class="s2">&quot;CCACTGCATGACTGGATTACAGGCA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CATGCCGGCGGTTAACGGATTACAC&quot;</span><span class="p">,</span> <span class="s2">&quot;CCGATTACAAGTGCTCTGCACGGCG&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CATATAGAGGCGATTACAGCGTATC&quot;</span><span class="p">,</span> <span class="s2">&quot;GGAACGATTACAGTGAGACTGCTCC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ATGATTACAGCGAAACGTATTCAAA&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTTCGGATGATTACACATTCTTCT&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GTACAATGCATCGCGATTACAACAC&quot;</span><span class="p">,</span> <span class="s2">&quot;GGATTACAAGTATCTGCCTGGATAC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTCCCGATTACATCAGGTACGTCCT&quot;</span><span class="p">,</span> <span class="s2">&quot;TAGAGAAGATTACAGCCTACTATTG&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AAGCTTTGGGCCGTACGATTACATC&quot;</span><span class="p">,</span> <span class="s2">&quot;GTAAGATTACAAGTTCAGGGTGATC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CATGATTACATTGGCGCCGACCTAC&quot;</span><span class="p">,</span> <span class="s2">&quot;GCTGGATTACAATCATACCCGTGTA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GGTTAGGGATTACAAACAAGACGTG&quot;</span><span class="p">,</span> <span class="s2">&quot;GACCGAGGTCTGATTACACTCCATC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ATAGACGCGATTACAAGCACTCTAA&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTCCGTTCTGCAGCTGATTACAAC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GGATTACACGCCTTCTCAAGCAGTG&quot;</span><span class="p">,</span> <span class="s2">&quot;ATCCTAACAGGATTACAAGAATTAC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;TATGAAGCTGAAGAAGATTACAGCA&quot;</span><span class="p">,</span> <span class="s2">&quot;CCTGTCTCAGATTACAGCACGGCGG&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="c1"># Initialize a new StruM object, using the DNA groove related</span>
<span class="c1"># features from the DiProDB table. Specify to use 4 cpus</span>
<span class="c1"># when doing EM.</span>
<span class="n">motif</span> <span class="o">=</span> <span class="n">strum</span><span class="o">.</span><span class="n">StruM</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;groove&#39;</span><span class="p">,</span> <span class="n">n_process</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="c1"># Train the model on the training sequences using expectation</span>
<span class="c1"># maximization, ensuring that that the variation of all</span>
<span class="c1"># position-specific features is at least 10e-5 (lim). Use</span>
<span class="c1"># a random_seed for reproducibility, and repeat with 20</span>
<span class="c1"># random restarts.</span>
<span class="n">k</span> <span class="o">=</span> <span class="mi">8</span>
<span class="n">motif</span><span class="o">.</span><span class="n">train_EM</span><span class="p">(</span><span class="n">training_sequences</span><span class="p">,</span> <span class="n">fasta</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">,</span>
    <span class="n">lim</span><span class="o">=</span><span class="mi">10</span><span class="o">**-</span><span class="mi">5</span><span class="p">,</span> <span class="n">random_seed</span><span class="o">=</span><span class="mi">620</span><span class="p">,</span> <span class="n">n_init</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="c1"># Examine the output, by identifying the best matching</span>
<span class="c1"># kmer in each of the training sequences.</span>
<span class="n">out</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">sequence</span> <span class="ow">in</span> <span class="n">training_sequences</span><span class="p">:</span>
    <span class="n">rseq</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">rev_comp</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="n">s1</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">score_seq</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
    <span class="n">s2</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">score_seq</span><span class="p">(</span><span class="n">rseq</span><span class="p">)</span>
    <span class="n">i1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">s1</span><span class="p">)</span>
    <span class="n">i2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">s2</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">s1</span><span class="p">[</span><span class="n">i1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">s2</span><span class="p">[</span><span class="n">i2</span><span class="p">]:</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">sequence</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">i1</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;+&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">rseq</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">i2</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>
    <span class="n">out</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">k</span><span class="p">])</span>
    <span class="nb">print</span> <span class="s2">&quot;</span><span class="si">{}{: &lt;2}</span><span class="s2"> {: &gt;</span><span class="si">{}</span><span class="s2">} </span><span class="si">{}</span><span class="s2"> </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">s</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">seq</span><span class="p">[:</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">(),</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="o">-</span><span class="n">k</span><span class="p">,</span>
        <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">(),</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="n">k</span><span class="p">:]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="p">)</span>

<span class="c1"># Summarize these best matches with a simple PWM.</span>
<span class="n">nucs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="s2">&quot;ACGT&quot;</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">)))</span>
<span class="n">PWM</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span><span class="n">k</span><span class="p">])</span>
<span class="k">for</span> <span class="n">thing</span> <span class="ow">in</span> <span class="n">out</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">n</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">thing</span><span class="p">):</span>
        <span class="n">PWM</span><span class="p">[</span><span class="n">nucs</span><span class="p">[</span><span class="n">n</span><span class="p">],</span> <span class="n">i</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

<span class="n">PWM</span> <span class="o">/=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">PWM</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">PWM</span><span class="p">:</span>
    <span class="nb">print</span> <span class="n">row</span>
</pre></div>
</div>
<p>There are several sections that are reported in the output. First, the StruM
gives a live update on how many iterations it took to converge for each of
the random restarts. Once they have all completed, it shows the likelihoods
for each of the restarts. The example output below highlights why it is
important to repeat this process a number of times, as the results are
highly variable. Finally we output a summary of the results. It is obvious
that the model correctly identified the sequence NGATTACA as observed both
in the highlighted sequences and the summary PWM at the bottom.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Retaining</span> <span class="mi">50</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">50</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">based</span> <span class="n">on</span> <span class="n">length</span> <span class="p">(</span><span class="o">&gt;</span><span class="mi">25</span><span class="n">bp</span><span class="p">)</span>
<span class="n">Detected</span> <span class="n">cyclical</span> <span class="n">likelihoods</span><span class="o">.</span> <span class="n">Proceeding</span> <span class="n">to</span> <span class="nb">max</span><span class="o">.</span>
<span class="n">Converged</span> <span class="n">after</span> <span class="mi">10</span> <span class="n">iterations</span> <span class="n">on</span> <span class="n">likelihood</span>
<span class="n">Converged</span> <span class="n">after</span> <span class="mi">5</span> <span class="n">iterations</span> <span class="n">on</span> <span class="n">likelihood</span>
<span class="o">...</span>
<span class="n">Restart</span> <span class="n">Likelihoods</span><span class="p">:</span> <span class="p">[</span><span class="mf">1981.3168691613439</span><span class="p">,</span> <span class="mf">1969.7041022864935</span><span class="p">,</span> <span class="mf">1969.7041022864935</span><span class="p">,</span> <span class="mf">650.89659719217332</span><span class="p">,</span> <span class="mf">386.24238189057303</span><span class="p">,</span> <span class="mf">369.62562708240256</span><span class="p">,</span> <span class="mf">350.54071957020324</span><span class="p">,</span> <span class="mf">350.54071957020324</span><span class="p">,</span> <span class="mf">350.54071957020324</span><span class="p">,</span> <span class="mf">327.8331936774872</span><span class="p">,</span> <span class="mf">315.95426601483587</span><span class="p">,</span> <span class="mf">242.73205512692093</span><span class="p">,</span> <span class="mf">235.90375348128853</span><span class="p">,</span> <span class="mf">149.92555284207441</span><span class="p">,</span> <span class="o">-</span><span class="mf">15.089679773020009</span><span class="p">,</span> <span class="o">-</span><span class="mf">68.060441249738645</span><span class="p">,</span> <span class="o">-</span><span class="mf">164.9202144304158</span><span class="p">,</span> <span class="o">-</span><span class="mf">373.49851273275601</span><span class="p">,</span> <span class="o">-</span><span class="mf">524.90598178843163</span><span class="p">,</span> <span class="o">-</span><span class="mf">524.91730067288267</span><span class="p">]</span>
<span class="o">+</span><span class="mi">0</span>                    <span class="n">CGATTACA</span> <span class="n">gatctcccgcgaccctt</span>
<span class="o">+</span><span class="mi">2</span>                 <span class="n">ga</span> <span class="n">TGATTACA</span> <span class="n">agatgcgtcgaatat</span>
<span class="o">+</span><span class="mi">12</span>      <span class="n">agccctgtccgc</span> <span class="n">AGATTACA</span> <span class="n">accac</span>
<span class="o">+</span><span class="mi">12</span>      <span class="n">agttaactccct</span> <span class="n">AGATTACA</span> <span class="n">tttgt</span>
<span class="o">...</span>
<span class="p">[</span> <span class="mf">0.26</span>  <span class="mf">0.</span>    <span class="mf">1.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">1.</span>    <span class="mf">0.</span>    <span class="mf">1.</span>  <span class="p">]</span>
<span class="p">[</span> <span class="mf">0.28</span>  <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">1.</span>    <span class="mf">0.</span>  <span class="p">]</span>
<span class="p">[</span> <span class="mf">0.24</span>  <span class="mf">1.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>  <span class="p">]</span>
<span class="p">[</span> <span class="mf">0.22</span>  <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">1.</span>    <span class="mf">1.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>    <span class="mf">0.</span>  <span class="p">]</span>
</pre></div>
</div>
</div>
<div class="section" id="adding-additional-features">
<h2><a class="toc-backref" href="#id4">Adding Additional Features</a><a class="headerlink" href="#adding-additional-features" title="Permalink to this headline">¶</a></h2>
<p>With the StruM package, in additional to the structural features
provided in the DiProDB table, you can incorporate additional arbitrary
features, as long as they provide a quantitative value across the
binding site.</p>
<p>In the example below we will define a function that looks up a DNase
signal from a <code class="docutils literal"><span class="pre">bigwig</span></code> file to incorporate. The file we will be using
comes from a DNase experiment in K562 cells, mapped to <em>hg19</em> from the
ENCODE project (<a class="reference external" href="https://www.encodeproject.org/files/ENCFF111KJD/">ENCFF111KJD</a>)
and can be downloaded from <a class="reference external" href="https://www.encodeproject.org/files/ENCFF111KJD/&#64;&#64;download/ENCFF111KJD.bigWig">here</a>.</p>
<p>The source code for this example can be found here: <a class="reference external" href="https://github.com/pdeford/StructuralMotifs/blob/master/examples/DNase.py">DNase.py</a>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Imports</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">bx.bbi.bigwig_file</span>
<span class="kn">from</span> <span class="nn">strum</span> <span class="k">import</span> <span class="n">strum</span>

<span class="c1"># Specify the path to where you downloaded the bigwig</span>
<span class="c1"># file. E.g. &quot;/Users/user/Downloads/ENCFF111KJD.bigWig&quot;</span>
<span class="n">DNase_bigwig_path</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

<span class="c1"># Define the function to be used by the StruM when</span>
<span class="c1"># converting from sequence-space to structural-space.</span>
<span class="c1"># NOTE: This function takes additional parameters.</span>
<span class="k">def</span> <span class="nf">lookup_DNase</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Lookup the signal in a bigWig file, convert NaNs to</span>
<span class="sd">    0s, ensure strandedness, and return the modified signal.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        data : (str) - Path to the bigWig file.</span>
<span class="sd">        chrom : (str) - Chromosome name.</span>
<span class="sd">        start : (int) - Base pair position on the chromsome of</span>
<span class="sd">                        the beginning of the region of interest.</span>
<span class="sd">        end : (int) - Base pair position on the chromsome of</span>
<span class="sd">                      the end of the region of interest.</span>

<span class="sd">    Returns:</span>
<span class="sd">        trace : (1D array) - The signal extracted from the</span>
<span class="sd">                             bigWig file for the region of</span>
<span class="sd">                             interest.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Open file</span>
    <span class="n">bwh</span> <span class="o">=</span> <span class="n">bx</span><span class="o">.</span><span class="n">bbi</span><span class="o">.</span><span class="n">bigwig_file</span><span class="o">.</span><span class="n">BigWigFile</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>
    <span class="c1"># Lookup signal for regions</span>
    <span class="n">trace</span> <span class="o">=</span> <span class="n">bwh</span><span class="o">.</span><span class="n">get_as_array</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># Clean up NaNs</span>
    <span class="n">trace</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">trace</span><span class="p">)]</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="c1"># Ensure strandedness</span>
    <span class="k">if</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">:</span>
        <span class="n">trace</span> <span class="o">=</span> <span class="n">trace</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">trace</span><span class="p">,</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># Some example sequences and their chromosome positions,</span>
<span class="c1"># from human build hg19.</span>
<span class="n">training_data</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">[</span><span class="s1">&#39;GAGATCCTGGGTTCGAATCCCAGC&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr6&#39;</span><span class="p">,</span> <span class="mi">26533165</span><span class="p">,</span> <span class="mi">26533141</span><span class="p">)],</span>
    <span class="p">[</span><span class="s1">&#39;GAGATCCTGGGTTCGAATCCCAGC&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="mi">33667901</span><span class="p">,</span> <span class="mi">33667925</span><span class="p">)],</span>
    <span class="p">[</span><span class="s1">&#39;GAGGTCCCGGGTTCGATCCCCAGC&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr6&#39;</span><span class="p">,</span> <span class="mi">28626033</span><span class="p">,</span> <span class="mi">28626009</span><span class="p">)],</span>
    <span class="p">[</span><span class="s1">&#39;GAGGTCCTGGGTTCGATCCCCAGT&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr6&#39;</span><span class="p">,</span> <span class="mi">28763754</span><span class="p">,</span> <span class="mi">28763730</span><span class="p">)],</span>
    <span class="p">[</span><span class="s1">&#39;GGGGGCGTGGGTTCGAATCCCACC&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr16&#39;</span><span class="p">,</span> <span class="mi">22308468</span><span class="p">,</span> <span class="mi">22308492</span><span class="p">)],</span>
    <span class="p">[</span><span class="s1">&#39;GGGGGCGTGGGTTCGAATCCCACC&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr5&#39;</span><span class="p">,</span> <span class="mi">180614647</span><span class="p">,</span> <span class="mi">180614671</span><span class="p">)],</span>
    <span class="p">[</span><span class="s1">&#39;AAGGTCCTGGGTTCGAGCCCCAGT&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr11&#39;</span><span class="p">,</span> <span class="mi">59318028</span><span class="p">,</span> <span class="mi">59318004</span><span class="p">)],</span>
    <span class="p">[</span><span class="s1">&#39;GAGGTCCCGGGTTCAAATCCCGGA&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">167684001</span><span class="p">,</span> <span class="mi">167684025</span><span class="p">)],</span>
    <span class="p">[</span><span class="s1">&#39;GAGGTCCCGGGTTCAAATCCCGGA&#39;</span><span class="p">,</span> <span class="p">(</span><span class="s1">&#39;chr7&#39;</span><span class="p">,</span> <span class="mi">128423440</span><span class="p">,</span> <span class="mi">128423464</span><span class="p">)],</span>
<span class="p">]</span>

<span class="c1"># Initialize a new StruM object.</span>
<span class="n">motif</span> <span class="o">=</span> <span class="n">strum</span><span class="o">.</span><span class="n">StruM</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;basic&#39;</span><span class="p">)</span>

<span class="nb">print</span> <span class="n">motif</span><span class="o">.</span><span class="n">features</span>

<span class="c1"># Update the StruM to incorporate the function</span>
<span class="c1"># defined above, drawing on the bigWig as the</span>
<span class="c1"># data source.</span>
<span class="n">motif</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">DNase_bigwig_path</span><span class="p">,</span> <span class="n">func</span><span class="o">=</span><span class="n">lookup_DNase</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;k562_DNase&#39;</span><span class="p">])</span>

<span class="nb">print</span> <span class="n">motif</span><span class="o">.</span><span class="n">features</span>

<span class="c1"># Train the model using the modified StruM and the example</span>
<span class="c1"># data above</span>
<span class="n">motif</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">training_data</span><span class="p">,</span> <span class="n">lim</span><span class="o">=</span><span class="mf">10.</span><span class="o">**-</span><span class="mi">5</span><span class="p">)</span>

<span class="c1"># Evaluate the similarity to the model of a new sequence.</span>
<span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;GAGGTCCCGGGTTCAATCCCCGGC&#39;</span>
<span class="n">position</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chr2&#39;</span><span class="p">,</span> <span class="mi">157257305</span><span class="p">,</span> <span class="mi">157257329</span><span class="p">]</span>

<span class="n">rseq</span> <span class="o">=</span> <span class="n">motif</span><span class="o">.</span><span class="n">rev_comp</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">rposition</span> <span class="o">=</span> <span class="p">[</span><span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">position</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">position</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>

<span class="n">s</span> <span class="o">=</span> <span class="p">[</span>
    <span class="n">motif</span><span class="o">.</span><span class="n">score_seq</span><span class="p">((</span><span class="n">seq</span><span class="p">,</span> <span class="n">position</span><span class="p">)),</span>
    <span class="n">motif</span><span class="o">.</span><span class="n">score_seq</span><span class="p">((</span><span class="n">rseq</span><span class="p">,</span> <span class="n">rposition</span><span class="p">))</span>
<span class="p">]</span>

<span class="n">strands</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">]</span>
<span class="nb">print</span> <span class="s2">&quot;Best match found on the &#39;</span><span class="si">{}</span><span class="s2">&#39; strand&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">strands</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">s</span><span class="p">)])</span>
</pre></div>
</div>
<p>This CTCF site is correctly identified as being in the forward
orientation, and there is an additional feature being considered.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="p">[</span><span class="s1">&#39;Twist&#39;</span><span class="p">,</span> <span class="s1">&#39;Rise&#39;</span><span class="p">,</span> <span class="s1">&#39;Bend&#39;</span><span class="p">]</span>
<span class="p">[</span><span class="s1">&#39;Twist&#39;</span><span class="p">,</span> <span class="s1">&#39;Rise&#39;</span><span class="p">,</span> <span class="s1">&#39;Bend&#39;</span><span class="p">,</span> <span class="s1">&#39;k562_DNase&#39;</span><span class="p">]</span>
<span class="n">Best</span> <span class="n">match</span> <span class="n">found</span> <span class="n">on</span> <span class="n">the</span> <span class="s1">&#39;+&#39;</span> <span class="n">strand</span>
</pre></div>
</div>
</div>
<div class="section" id="plotting-a-graphical-representation-of-the-strum">
<h2><a class="toc-backref" href="#id5">Plotting a graphical representation of the StruM</a><a class="headerlink" href="#plotting-a-graphical-representation-of-the-strum" title="Permalink to this headline">¶</a></h2>
<p>You may find it useful to look at a graphical representation of the
motif learned using the methods above. This can be accomplished useing
the <a class="reference internal" href="code.html#strum.StruM.plot" title="strum.StruM.plot"><code class="xref py py-func docutils literal"><span class="pre">strum.StruM.plot()</span></code></a> method. Each of the features included in
the StruM will be displayed as its own linegraph. The line represents
the average value of that feature across the site, and the shaded area
represents +/- 1 standard deviation.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Imports</span>
<span class="kn">from</span> <span class="nn">strum</span> <span class="k">import</span> <span class="n">strum</span>

<span class="c1"># Sequences representing some of the variability of the</span>
<span class="c1"># FOXA1 binding site.</span>
<span class="n">training_sequences</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;CTGTGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTAAGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTGTGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTGTGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CAGAGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTAAGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTGTGCAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CAATGTAAACA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CTGAGTAAATA&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="c1"># Train the motif using the &#39;groove&#39; related features</span>
<span class="n">motif</span> <span class="o">=</span> <span class="n">strum</span><span class="o">.</span><span class="n">StruM</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;groove&#39;</span><span class="p">)</span>
<span class="n">motif</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">training_sequences</span><span class="p">,</span> <span class="n">fasta</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Plot the the StruM, and save to the specified filename</span>
<span class="n">motif</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="s1">&#39;FOXA1_strum.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>This code produces the image below. If you consider the Major Groove
Distance, you will notice that there is more variation at the
beginning of the motif (to the left) than at the end, as indicated by
the shaded region. It is also clear to see that FOXA1 prefers a higher
value at position 6, and a low value at position 9.</p>
<img alt="_images/strumplot.png" src="_images/strumplot.png" />
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="code.html" class="btn btn-neutral float-right" title="Documentation" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral" title="Installation Instructions" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Peter DeFord.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>